openupgrade_scripts/scripts/microsoft_calendar/18.0.1.0/pre-migration.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2025 L4 Tech S.L.
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from openupgradelib import openupgrade
+
+
+def calendar_precompute_ms_universal_event_id(env):
+    """
+    Now in 18.0 the ms_universal_event_id is no longer computed from microsoft_id
+    - https://github.com/odoo/odoo/blob/18.0/addons/microsoft_calendar/models/microsoft_sync.py#L411
+    Now the id is defined directly from the Microsoft API call, more
+    specifically field 'iCalUId'
+    - https://github.com/odoo/odoo/blob/18.0/addons/microsoft_calendar/utils/microsoft_calendar.py#L159
+    So, in order for past events to have a ms_universal_event_id we
+    must precompute by splitting by ':' the microsoft_id
+    - https://github.com/odoo/odoo/blob/18.0/addons/microsoft_calendar/models/microsoft_sync.py#L120
+    """
+
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE calendar_recurrence
+        SET ms_universal_event_id = split_part(microsoft_id, ':', 2)
+        WHERE microsoft_id LIKE '%:%';
+        """,
+    )
+
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE calendar_event
+        SET ms_universal_event_id = split_part(microsoft_id, ':', 2)
+        WHERE microsoft_id LIKE '%:%';
+        """,
+    )
+
+
+def microsoft_calendar_credentials_to_res_users_settings(env):
+    """
+    microsoft.calendar.credentials has been deleted and now settings are
+    stored in res.users.settings. For each user in microsoft.calendar.credentials
+    user_ids (one2many) field we must copy into that user's res.users.settings
+    (user_id, many2one) the following fields:
+    * calendar_sync_token -> microsoft_calendar_sync_token
+    * last_sync_date -> microsoft_last_sync_date
+    * synchronization_stopped -> microsoft_synchronization_stopped
+    """
+
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE res_users_settings AS rus
+        SET
+            rus.microsoft_calendar_sync_token = cred.calendar_sync_token,
+            rus.microsoft_last_sync_date = cred.last_sync_date,
+            rus.microsoft_synchronization_stopped = cred.synchronization_stopped
```

Comment:
```suggestion
            microsoft_calendar_sync_token = cred.calendar_sync_token,
            microsoft_last_sync_date = cred.last_sync_date,
            microsoft_synchronization_stopped = cred.synchronization_stopped
```
In `SET` command, never put table.